plugins/linux/iptablesplugin/descriptor/rulechain.go

```diff
@@ -0,0 +1,418 @@
+			return nil, err
+		}
+	}
+
+	// wipe all rules in the chain that may have existed before
+	err = d.ipTablesHandler.DeleteAllRules(protocolType(rch), tableNameStr(rch), chainNameStr(rch))
+	if err != nil {
+		d.log.Warnf("Error by wiping iptables rules: %v", err)
+	}
+
+	// append all rules
+	for _, rule := range rch.Rules {
+		err := d.ipTablesHandler.AppendRule(protocolType(rch), tableNameStr(rch), chainNameStr(rch), rule)
+		if err != nil {
+			d.log.Errorf("Error by appending iptables rule: %v", err)
+			break
+		}
+	}
+
+	return nil, err
+}
+
+// Delete removes iptables rule chain.
+func (d *RuleChainDescriptor) Delete(key string, rch *linux_iptables.RuleChain, metadata interface{}) error {
+
+	d.log.Debugf("DELETE IPT rule chain %s: %v", key, rch)
+
+	// switch network namespace
+	nsCtx := nslinuxcalls.NewNamespaceMgmtCtx()
+	nsRevert, err := d.nsPlugin.SwitchToNamespace(nsCtx, rch.Namespace)
+	if err != nil {
+		d.log.WithFields(logging.Fields{
+			"err":       err,
+			"namespace": rch.Namespace,
+		}).Warn("Failed to switch the namespace")
+		return err
+	}
+	// revert network namespace after returning
+	defer nsRevert()
+
+	// delete all rules in the chain
+	err = d.ipTablesHandler.DeleteAllRules(protocolType(rch), tableNameStr(rch), chainNameStr(rch))
+	if err != nil {
+		d.log.Errorf("Error by deleting iptables rules: %v", err)
+	}
+
+	// delete the chain if it was custom-defined
+	if rch.ChainType == linux_iptables.RuleChain_CUSTOM {
+		err := d.ipTablesHandler.DeleteChain(protocolType(rch), tableNameStr(rch), chainNameStr(rch))
+		if err != nil {
+			d.log.Errorf("Error by deleting iptables chain: %v", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// UpdateWithRecreate tells the scheduler that the update should be performed with re-create.
+func (d *RuleChainDescriptor) UpdateWithRecreate(key string, oldRCh, newRCh *linux_iptables.RuleChain, oldMetadata interface{}) bool {
+	// always update with recreate
+	return true
+}
+
+// Dependencies lists dependencies for a iptables rule chain.
+func (d *RuleChainDescriptor) Dependencies(key string, rch *linux_iptables.RuleChain) []kvs.Dependency {
+	if len(rch.Interfaces) > 0 {
+		// the associated interfaces must exist
+		var deps []kvs.Dependency
+		for _, i := range rch.Interfaces {
+			deps = append(deps, kvs.Dependency{
+				Label: ruleChainInterfaceDep,
+				Key:   ifmodel.InterfaceKey(i),
+			})
+		}
+		return deps
+	}
+	return nil
+}
+
+// retrievedRuleChains is used as the return value sent via channel by retrieveRuleChains().
+type retrievedRuleChains struct {
+	chains []adapter.RuleChainKVWithMetadata
+	err    error
+}
+
+// Retrieve returns all iptables rule chain entries managed by this agent.
+func (d *RuleChainDescriptor) Retrieve(correlate []adapter.RuleChainKVWithMetadata) ([]adapter.RuleChainKVWithMetadata, error) {
+	var values []adapter.RuleChainKVWithMetadata
+
+	goRoutinesCnt := len(correlate) / minWorkForGoRoutine
```

Comment:
Retrieve every namespace only once or there will be duplicities if multiple chains for the same namespace are in correlate, e.g.: https://github.com/ligato/vpp-agent/blob/master/plugins/linux/ifplugin/descriptor/interface.go#L613-L621